### PR TITLE
versioning list object versions

### PR DESCRIPTION
**Pull request by rhsakarpos**
**this PR adds the ability to ListObjects to return all versions of all objects**:

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

**Release note**:

```

1. Support listing all versions of all objects
2. Instead of returning a list of objects, now returns a list of list of objects

```

### CODE REVIEW

**Comment by rhsakarpos** [comment]
@kumarashit @wisererik @skdwriting @anvithks @sfzeng @nguptaopensds please review, thanks

File: api/pkg/s3/bucketget.go [hunk 1]
```diff
@@ -145,30 +155,18 @@ func CreateListObjectsResponse(bucketName string, request *s3.ListObjectsRequest
 		response.Contents = append(response.Contents, obj)
 	}
 
-	var prefixes []datatype.CommonPrefix
```
Comment (sfzeng) [inline api/pkg/s3/bucketget.go]: Hi @rhsakarpos is it compatible with aws s3 if remove those codes?

**Comment by sfzeng** [comment]
Hi @rhsakarpos What do you mean by 'Instead of returning a list of objects, now returns a list of list of objects'? Is it compatible with aws s3?
BTW, fix the CI error, please. Thanks.

**Comment by skdwriting** [comment]
@rhsakarpos build errors

File: api/pkg/s3/bucketget.go (hunk 1 above)
Comment (rhsakarpos) [reply api/pkg/s3/bucketget.go]: hi @sfzeng i need to check the AWS spec and align this(with support for all cloud backends) as needed, since now, we are NOT setting versioning on the cloud backend. I plan to do this after RC1, does that work ? I'll raise an issue to track it for RC2 ?

File: api/cmd/main.go [hunk 2]
```diff
@@ -62,7 +61,7 @@ func main() {
 
 	s3ws.Filter(logging.FilterFactory())
 	s3ws.Filter(context.FilterFactory())
-	s3ws.Filter(signer.FilterFactory())
+	//s3ws.Filter(signer.FilterFactory())
```
Comment (nguptaopensds) [inline api/cmd/main.go]: commenting this line was a part of testing we do in our local environment, this should not be committed

File: api/pkg/policy/policy.go [hunk 3]
```diff
@@ -203,10 +203,11 @@ func Authorize(req *restful.Request, res *restful.Response, action string) bool
 	log.Infof("Action: %v", action)
 	log.Infof("Target: %v", target)
 	log.Infof("policy-Credentials: %v", credentials)
-	ok, err := enforcer.Authorize(action, target, credentials)
+	ok := true
+	/*ok, err := enforcer.Authorize(action, target, credentials)
```
Comment (nguptaopensds) [inline api/pkg/policy/policy.go]: Please add comment for why is this needed

File: s3/pkg/meta/db/drivers/tidb/bucket.go [hunk 4]
```diff
@@ -313,11 +313,11 @@ func (t *TidbClient) ListObjects(ctx context.Context, bucketName string, version
 			objectNum[name] += 1
 			filter[common.KMarker] = name
 
-			if _, ok := objectMap[name]; !ok {
+			/*if _, ok := objectMap[name]; !ok {
```
Comment (nguptaopensds) [inline s3/pkg/meta/db/drivers/tidb/bucket.go]: please remove commented code if not required

File: s3/pkg/meta/db/drivers/tidb/object.go [hunk 5]
```diff
@@ -48,10 +48,10 @@ func (t *TidbClient) GetObject(ctx context.Context, bucketName, objectName, vers
 	err = row.Scan(
 		&ibucketname,
 		&iname,
-		&iversion,
+		&object.VersionId,
 		&object.Location,
-		&object.TenantId,
-		&object.UserId,
+		//&object.TenantId,
```
Comment (nguptaopensds) [inline s3/pkg/meta/db/drivers/tidb/object.go]: Can you put some comment as to why we have removed tenant and userId from here, will this affect Multi-tenant support in multi-cloud?

File: s3/pkg/service/object.go [hunk 6]
```diff
@@ -552,8 +564,39 @@ func (s *s3Service) ListObjects(ctx context.Context, in *pb.ListObjectsRequest,
...
+			// add a new obj list
+			objList = pb.ListObjects{
+				Objects:              make([]*pb.Object, 0),
+				XXX_NoUnkeyedLiteral: struct{}{},
+				XXX_unrecognized:     nil,
+				XXX_sizecache:        0,
+			}
+			objList.Objects = append(objList.Objects, &object)
+			mapNameToListOfObjects[object.ObjectKey] = objList
+		}else{
```
Comment (nguptaopensds) [inline s3/pkg/service/object.go]: Can you explain this else part

**Comment by sfzeng** [comment]
Hi @rhsakarpos Could you please reply to the comments above and fix conflict? Thanks!

**Comment by rhsakarpos** [comment]
> Hi @rhsakarpos Could you please reply to the comments above and fix conflict? Thanks!

thanks, will do. I am re-raising the PR with the comments addressed, will update when done, ETA an hour or so

**Comment by rhsakarpos** [comment]
re-raised https://github.com/opensds/multi-cloud/pull/779
#779
